x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java

```diff
@@ -32,6 +32,91 @@
 import static org.elasticsearch.xpack.esql.core.util.PlanStreamInput.readCachedStringWithVersionCheck;
 import static org.elasticsearch.xpack.esql.core.util.PlanStreamOutput.writeCachedStringWithVersionCheck;
 
+/**
+ * This enum represents data types the ES|QL query processing layer is able to
+ * interact with in some way. This includes fully representable types (e.g.
+ * {@link DataType#LONG}, numeric types which we promote (e.g. {@link DataType#SHORT})
+ * or fold into other types (e.g. {@link DataType#DATE_PERIOD}) early in the
+ * processing pipeline, types for internal use
+ * cases (e.g. {@link DataType#PARTIAL_AGG}), and types which the language
+ * doesn't support, but require special handling anyway (e.g.
+ * {@link DataType#OBJECT})
+ *
+ * <h2>Process for adding a new data type</h2>
```

Comment:
Just a comment: Should this be here in the class javadoc? I think we have "how to's" all spread around different places (package-infos, docs?, now classes).
Better here than nowhere I guess, but it feels strange to have this kind of docs in a class

Reply:
I'm open to suggestions for a better place to put it.  This seemed the most findable to me at the time.

Reply:
Maybe we should link it from the main esql package-info docs? So it will be easier to find, and we won't forget about it